### PR TITLE
DigitalOcean DNS Authenticator

### LINKED ISSUE #4570: Implement a DNS authenticator for DigitalOcean
Implement an Authenticator plugin which can fulfill a dns-01 challenge using the DigitalOcean API.

See #4423.

### PR DESCRIPTION
**Pull request by zjs**
Implement an Authenticator which can fulfill a dns-01 challenge using the
DigitalOcean API. Applicable only for domains using DigitalOcean for DNS.

Testing Done:
 * `tox -e py27`
 * `tox -e lint`
 * Manual testing:
    * Used `certbot certonly --dns-digitalocean -d`, specifying a
      credentials file as a command line argument. Verified that a
      certificate was successfully obtained without user interaction.
    * Used `certbot certonly --dns-digitalocean -d`, without specifying a
      credentials file as a command line argument. Verified that the user
      was prompted and that a certificate was successfully obtained.
    * Used `certbot certonly -d`. Verified that the user was prompted for
      a credentials file after selecting digitalocean interactively and
      that a certificate was successfully obtained.
    * Used `certbot renew --force-renewal`. Verified that certificates
      were renewed without user interaction.
 * Negative testing:
    * Path to non-existent credentials file.
    * Credentials file with unsafe permissions (644).
    * Credentials file missing token.
    * Credentials file with blank token.
    * Credentials file with incorrect token.
    * Domain name not registered to DigitalOcean account.
 * `./tox.cover.sh`
    * certbot_dns_digitalocean/dns_digitalocean.py            97%   58, 85
    * certbot_dns_digitalocean/dns_digitalocean_test.py     100%

(Replaces #4423, resolves #4570)

### CODE REVIEW

File: certbot-dns-digitalocean/certbot_dns_digitalocean/dns_digitalocean.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                logger.debug('Found base domain for %s using name %s', domain_name, guess)
+                return domain
+
+        raise errors.PluginError('Unable to determine base domain for {0} using names: {1}.'
+                                 .format(domain_name, domain_name_guesses))
+
+    @staticmethod
+    def _compute_record_name(domain, full_record_name):
+        # The domain, from DigitalOcean's point of view, is automatically appended.
+        return full_record_name.split("." + domain.name)[0]
```
Comment (bmw) [inline certbot-dns-digitalocean/certbot_dns_digitalocean/dns_digitalocean.py]: While silly, if the user attempts to validate a name containing the domain from Digital Ocean's point of view multiple times this doesn't work. For example, Digital Ocean sees my domain as `example.org`, but if I attempt to get a cert for `example.org.example.org` I fail the challenge.

I believe we can work around the problem by changing this to:
```
return full_record_name.rpartition("." + domain.name)[0]
```

**Comment by glenjamin** [comment]
This looks ace, but is there any documentation about how to use it? Is it installed by default?

**Comment by bmw** [comment]
Unfortunately, this plugin is not installed by default and installing it through our currently supported installation methods is pretty difficult (unless you're running Arch Linux in which case you can run `sudo pacman -S certbot-dns-digitalocean`). We're changing this in preparation for [wildcard certs](https://letsencrypt.org/2017/12/07/looking-forward-to-2018.html#new-features), but in the meantime, the best way to use this is:

1. [Download certbot-auto](https://certbot.eff.org/docs/install.html#certbot-auto).
2. Run `./certbot-auto --os-packages-only`
3. Run `virtualenv /path/you/want/to/install/certbot`
4. Run `/path/you/want/to/install/certbot/bin/pip install certbot-dns-digitalocean`

After that, you can run Certbot with `/path/you/want/to/install/certbot/bin/certbot` and upgrade to a newer version with `/path/you/want/to/install/certbot/bin/pip install -U certbot-dns-digitalocean`.

You can get more information on how to use the plugin from `/path/you/want/to/install/certbot/bin/certbot --help dns-digitalocean` and [this file](https://github.com/certbot/certbot/blob/master/certbot-dns-digitalocean/certbot_dns_digitalocean/__init__.py).

**Comment by glenjamin** [comment]
After posting my question I had a go at this and managed to get it working.

On Ubuntu 16.04, here's the ansible playbook I used

```yaml
- name: add certbot repo
  apt_repository:
    repo: ppa:certbot/certbot

- name: install packages
  apt:
    pkg: "{{item}}"
  with_items:
    - certbot
    - python3-pip
    # Dependencies of certbot-dns-digitalocean
    # these appear to be missing or broken if installed via pip
    - python3-openssl
    - python3-jsonpickle
    - python3-ndg-httpsclient

- name: install digitialocean plugin
  pip:
    name: certbot-dns-digitalocean
    version: 0.20.0

- name: certbot config file
  copy:
    dest: /etc/letsencrypt/cli.ini
    content: |
      {% for k, v in certbot_options.iteritems() %}
      {{ k }}={{ v }}
      {% endfor %}
  vars:
    certbot_options:
      email: <blah>
      no-eff-email: true
      authenticator: dns-digitalocean
      agree-tos: true
      non-interactive: true
      expand: true
      dns-digitalocean-credentials: /etc/letsencrypt/digitalocean.ini
      staging: "{{ certbot_staging }}"

- name: digitalocean credentials
  copy:
    dest: /etc/letsencrypt/digitalocean.ini
    mode: 0600
    content: |
      dns_digitalocean_token={{ lookup('env', 'DIGITALOCEAN_DNS_TOKEN') }}
```

**Pull request merged**